Cabal/Distribution/Compat/Async.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE CPP                #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+-- | 'Async', yet using 'MVar's.
+--
+-- Adopted from @async@ library
+-- Copyright (c) 2012, Simon Marlow
+-- Licensed under BSD-3-Clause
+--
+module Distribution.Compat.Async (
+    AsyncM,
+    withAsync, waitCatch,
+    wait, asyncThreadId,
+    cancel, uninterruptibleCancel, AsyncCancelled (..),
+    ) where
+
+import Control.Concurrent      (ThreadId, forkIO)
+import Control.Concurrent.MVar (MVar, newEmptyMVar, putMVar, readMVar)
+import Control.Exception
+       (BlockedIndefinitelyOnMVar (..), Exception (..), SomeException (..), catch, mask, throwIO,
+       throwTo, try, uninterruptibleMask_)
+import Control.Monad           (void)
+import Data.Typeable           (Typeable)
+import GHC.Exts                (inline)
+
+-- TODO: base?
+#if __GLASGOW_HASKELL__ >= 708
+import Control.Exception (asyncExceptionFromException, asyncExceptionToException)
+#endif
+
+-- | Async, but based on 'MVar', as we don't depend on @stm@.
+data AsyncM a = Async
```

Comment:
What's the motivation for not just naming it `Async`? Isn't the MVar vs STM thing merely an internal implementation detail that doesn't affect the public API?

Reply:
To make it clear that it's different.

Reply:
... in some sense, I'd like to go over the use sites if/when we change to STM based implementation.